Better mapper cache interface

glightfoot opened the pull request:
Kind of fixes #295 and kind of fixes #362 by removing the lru code from the mapper package. 

First stab at moving this to a better interface. The noop cache is now default (was lru), and the rr and lru mapper caches have been moved to a different package. This makes it easier for users of the packages to implement their own caches as the interface is much simpler.

Instead of calling InitCache, there is a UseCache method on mapper to set a cache.

I still want to rework the metrics, but need to think about a better way to do so. I wanted to get some feedback before spending any more time on it.

@matthiasr 

Signed-off-by: glightfoot <glightfoot@rsglab.com>

glightfoot commented:
I just now realized the mapper cache issue was actually to wrap the mapper with a cache. I didn’t actually read that issue in detail, but I think having the mapper use a cache is important enough of a performance improvement to keep in there. Thoughts?

matthiasr commented on pkg/mapper/mapper.go:
```diff
@@ -252,53 +256,34 @@ func (m *MetricMapper) InitFromYAMLString(fileContents string, cacheSize int, op
-	result, cached := m.cache.Get(statsdMetric, statsdMetricType)
+
+	// default cache to noop cache if used from an uninitialized mapper
+	if m.cache == nil {
+		m.cache = NewMetricMapperNoopCache()
+	}
```
I think we need to take an exclusive lock here, not just an RLock?

matthiasr replied:
I guess for performance _and_ correctness we need to check twice? (concurrency is not my strength)

```suggestion
	if m.cache == nil {
		m.mutex.Lock()
		if m.cache == nil {
			m.cache = NewMetricMapperNoopCache()
		}
		m.mutex.Unlock()
	}
```

glightfoot replied:
good catch! yeah, we will need to check twice

glightfoot replied:
It might make more sense to remove the noop cache entirely, and just hide the part that checks the cache for a result behind the `if m.cache == nil`

matthiasr replied:
aaawwwe I liked it (again, aesthetically) but I don't feel strongly about this

matthiasr commented on pkg/mapper/mapper_benchmark_test.go:
```diff
@@ -582,9 +585,18 @@ mappings:` + duplicateRules(100, ruleTemplateSingleMatchGlob)
 	}
 
 	for _, cacheType := range []string{"lru", "random"} {
+		mapper := MetricMapper{}
+		var cache MetricMapperCache
+		switch cacheType {
+		case "lru":
+			cache, _ = lru.NewMetricMapperLRUCache(mapper.Registerer, 1000)
+		case "random":
+			cache, _ = randomreplacement.NewMetricMapperRRCache(mapper.Registerer, 1000)
+		}
```
a bunch of tests are repeating this. would it be easier to have a helper that returns a (fresh) `map[string]MetricMapperCache` so we can `range cacheType, cache := cacheTypes() { … }`?

matthiasr replied:
could we generalize things if that also includes the `noop` cache?

glightfoot replied:
yeah, I'll add a helper function for this.

matthiasr commented on pkg/mapper/mapper_cache.go:
```diff
@@ -56,155 +53,36 @@ func NewCacheMetrics(reg prometheus.Registerer) *CacheMetrics {
-	AddMatch(metricString string, metricType MetricType, mapping *MetricMapping, labels prometheus.Labels)
-	AddMiss(metricString string, metricType MetricType)
+	// Get a cached result
+	Get(metricKey string) (interface{}, bool)
```
What is the benefit of this compared to keeping the metricString and metricType separate?

glightfoot replied:
In my mind, it made more sense to build the key from the two parts and pass it into the cache already formatted. This way it makes the cache interface more generic (key-value, vs key1 key2 - value), and also keeps people from having to duplicate the key-building code in each cache package.

matthiasr reviewed (commented):
Aesthetically I would find the wrapping approach more pleasing, but this is definitely an improvement, and I am happy with it! I left a few comments but nothing major stands out.

glightfoot commented:
@matthiasr I removed the noop cache entirely and hid the cache calls behind `if m.cache != nil` checks, which seems simpler. Also added a helper function for the tests

matthiasr reviewed (approved):
LGTM! What is missing to take it out of Draft?

glightfoot commented:
Ahh sorry about that! Got busy with a job change. Should be ready for review

matthiasr commented:
Thanks a lot!
